File: src/main/java/org/vanautrui/languages/parsing/astnodes/nonterminal/OperatorNode.java
package org.vanautrui.languages.parsing.astnodes.nonterminal;

import org.vanautrui.languages.lexing.collections.TokenList;
import org.vanautrui.languages.lexing.tokens.utils.Token;
import org.vanautrui.languages.lexing.tokens.OperatorToken;
import org.vanautrui.languages.parsing.IASTNode;

public class OperatorNode implements IASTNode {

    //same as the token

    public String operator;

    public OperatorNode(TokenList tokens)throws Exception{

        TokenList copy = new TokenList(tokens);

        Token token = copy.get(0);

        if(token instanceof OperatorToken){
            this.operator=((OperatorToken)token).operator;
            copy.consume(1);
        }else{
            throw new Exception("could not recognize operator token: got "+token.getContents());
        }

        tokens.set(copy);
    }

    @Override
    public String toSourceCode() {
        return " "+operator+" ";
    }
}

File: dragon-parser/src/main/parsing/OperatorNode.cpp
#include <iostream>
#include <vector>
#include <string>
#include "OperatorNode.hpp"
#include "../commandline/TokenList.hpp"
#include "../commandline/BaseToken.hpp"
#include "../commandline/TokenKeys.hpp"

using namespace std;


OperatorNode::OperatorNode(TokenList tokens, bool debug) {

	if(debug){
		cout << "OperatorNode(...)" << endl;
	}

	TokenList copy = TokenList(tokens);

	BaseToken token = copy.get(0);

	if (token.kind == OPERATOR) {

		if (copy.get(1).kind == OPERATOR) {
			BaseToken opleft = token;
			BaseToken opright = copy.get(1);

			if ((opleft.value.compare("<")==0) || opleft.value.compare(">")==0
					&& opright.value.compare("=")==0
			) {
				this->op = opleft.value + opright.value;
				copy.consume(2);
			} else if (opleft.value.compare("=")==0 && opright.value.compare("=")==0) {
				this->op = "==";
				copy.consume(2);
			} else {
				throw "could not recognize operator token: got " + token.value;
			}
		} else {
			this->op = (token).value;
			copy.consume(1);
		}
	} else {
		throw "could not recognize operator token: got " + token.value;
	}

	tokens.set(copy);
}

